docs/RFCS/20201020_sql_query_denylist.md

```diff
@@ -0,0 +1,183 @@
+
+# Guide-level explanation
+
+## New named concepts
+
+### Denylist
+The denylist we define for this RFC and plan to implement is a global, non-configurable SQL query denylist in CockroachDB. A location for this denylist will be configured via a CLI flag, and if a denylist is uploaded to that location, any following queries matching any pattern on the denylist file will be denied. The supported file format for the denylist is YAML; each query pattern on the denylist file must be specified via regular expression conventions. 
+
+### Staging denylist
+The staging denylist is intended to be used so that a user will be able to test whether the query patterns on their denylist deny the appropriate queries, without actually blocking them. A separate location for the staging denylist will be configured via a CLI flag. Otherwise, it should be formatted in the same way as the denylist. 
+
+## Examples
+
+### User experience from a database administrator's point of view
+A database administrator, anticipating that a denylist or staging denylist may be used in the future, uses a flag to configure the location where the denylist would be located when it is needed. When they notice a cluster is unresponsive, and isolate the root cause to a particular query (or multiple), they optionally may want to confirm the regex query pattern matches the queries they want to deny. 
+
+To confirm, they can add a staging denylist to the location they had specified for the staging denylist and execute the query. The query will execute, and if the query matches the regex query pattern specified in the denylist, the administrator will be able to view this in the Admin UI, SQL console, and CRDB logs. 
+
+The administrator will then be able to add a denylist to the denylist location specified. Any queries after this moment will be matched against the denylist, and the denial logged for the administrator if there is a match. A time-series graph of queries denied, and queries that would be denied in the case of a staging denylist, will also be available to view on the Admin UI.
+
+The administrator may modify the file content as needed; and if the entire denylist is no longer available, they may remove it entirely. The changes will take effect immediately.
+
+### User experience from a developer's point of view
+A developer may be sending a query that has been denied on the denylist or would be denied and is currently on the staging denylist. In the latter case, they will also be able to view the warning on console that this query is still executing, but would be denied due to a pattern match on a staging denylist. In the former case, they will receive an error message that "query matched a pattern in the denylist by the database administrator" with the specific regex query pattern that it matched. 
```

Comment:
`by the database administrator` could be misleading since anyone with filesystem write access could modify the denylist. We should include the denylist location?

Reply:
Yeah, that's fair. I don't think there's a problem with exposing the denylist location to the application developer.